public/app/plugins/datasource/loki/add_label_to_query.ts

```diff
@@ -1,140 +1,161 @@
-import { chain, isEqual } from 'lodash';
-
-import { PROM_KEYWORDS, OPERATORS, LOGICAL_OPERATORS } from 'app/plugins/datasource/prometheus/promql';
-
-import { LOKI_KEYWORDS } from './syntax';
-
-const builtInWords = [...PROM_KEYWORDS, ...OPERATORS, ...LOGICAL_OPERATORS, ...LOKI_KEYWORDS];
-
-// We want to extract all possible metrics and also keywords
-const metricsAndKeywordsRegexp = /([A-Za-z:][\w:]*)\b(?![\]{=!",])/g;
-
-export function addLabelToQuery(
-  query: string,
-  key: string,
-  value: string | number,
-  operator?: string,
-  hasNoMetrics?: boolean
-): string {
+import { parser } from '@grafana/lezer-logql';
+
+import { QueryBuilderLabelFilter } from '../prometheus/querybuilder/shared/types';
+
+import { LokiQueryModeller } from './querybuilder/LokiQueryModeller';
+import { buildVisualQueryFromString } from './querybuilder/parsing';
+import { LokiVisualQuery } from './querybuilder/types';
+
+/**
+ * Adds label filter to existing query. Useful for query modification for example for ad hoc filters.
+ *
+ * It uses LogQL parser to find instances of labels, alters them and then splices them back into the query.
+ * In a case when we have parser, instead of adding new instance of label it adds label filter after the parser.
+ *
+ * This operates on substrings of the query with labels and operates just on those. This makes this
+ * more robust and can alter even invalid queries, and preserves in general the query structure and whitespace.
+ *
+ * @param query
+ * @param key
+ * @param value
+ * @param operator
+ */
+export function addLabelToQuery(query: string, key: string, value: string | number, operator = '='): string {
```

Comment:
operator needs to be string, because in ad-hoc filters, it possibly can have different value (e.g. `>`,...). This is not really usable in Loki, but you can select.

Reply:
Changed typing here:  https://github.com/grafana/grafana/pull/51232/commits/f5d09970884af3bce994473d486afe305ed257f6

Changed order here: https://github.com/grafana/grafana/pull/51232/commits/23df85cc83f7eafcd08f369e36c2db79853b0f01